File: backend/models/note.py
"""
Note Model for NotesVault Backend
Handles note creation, storage, and management
"""

from app import db
from datetime import datetime
from sqlalchemy.orm import relationship

class Note(db.Model):
    """Note model for storing and managing student notes"""

    __tablename__ = 'notes'

    id = db.Column(db.Integer, primary_key=True)
    title = db.Column(db.String(200), nullable=False, index=True)
    content = db.Column(db.Text, nullable=False)
    subject = db.Column(db.String(100), nullable=False, index=True)
    tags = db.Column(db.String(500), nullable=True)  # Comma-separated tags
    is_public = db.Column(db.Boolean, default=True, nullable=False)
    is_featured = db.Column(db.Boolean, default=False, nullable=False)
    views_count = db.Column(db.Integer, default=0, nullable=False)
    likes_count = db.Column(db.Integer, default=0, nullable=False)
    created_at = db.Column(db.DateTime, default=datetime.utcnow, nullable=False)
    updated_at = db.Column(db.DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)

    # Foreign Key
    user_id = db.Column(db.Integer, db.ForeignKey('users.id'), nullable=False, index=True)

    # Relationships
    user = relationship('User', back_populates='notes')

    def __init__(self, title, content, subject, user_id, tags=None, is_public=True):
        self.title = title
        self.content = content
        self.subject = subject
        self.user_id = user_id
        self.tags = tags
        self.is_public = is_public

    def __repr__(self):
        return f'<Note {self.title}>'

    def to_dict(self, include_content=True):
        """Convert note object to dictionary"""
        note_dict = {
            'id': self.id,
            'title': self.title,
            'subject': self.subject,
            'tags': self.get_tags_list(),
            'is_public': self.is_public,
            'is_featured': self.is_featured,
            'views_count': self.views_count,
            'likes_count': self.likes_count,
            'created_at': self.created_at.isoformat() if self.created_at else None,
            'updated_at': self.updated_at.isoformat() if self.updated_at else None,
            'user_id': self.user_id,
            'author': self.user.username if self.user else None
        }

        if include_content:
            note_dict['content'] = self.content

        return note_dict

    def get_tags_list(self):
        """Get tags as a list instead of comma-separated string"""
        if self.tags:
            return [tag.strip() for tag in self.tags.split(',') if tag.strip()]
        return []

    def set_tags_from_list(self, tags_list):
        """Set tags from a list"""
        if tags_list:
            self.tags = ','.join([tag.strip() for tag in tags_list if tag.strip()])
        else:
            self.tags = None

    def increment_views(self):
        """Increment view count"""
        self.views_count += 1
        db.session.commit()

    def increment_likes(self):
        """Increment like count"""
        self.likes_count += 1
        db.session.commit()

    def decrement_likes(self):
        """Decrement like count (minimum 0)"""
        if self.likes_count > 0:
            self.likes_count -= 1
            db.session.commit()

    @classmethod
    def get_public_notes(cls, limit=None, subject=None, search_term=None):
        """Get public notes with optional filtering"""
        query = cls.query.filter_by(is_public=True)

        if subject:
            query = query.filter_by(subject=subject)

        if search_term:
            search_pattern = f'%{search_term}%'
            query = query.filter(
                db.or_(
                    cls.title.ilike(search_pattern),
                    cls.content.ilike(search_pattern),
                    cls.tags.ilike(search_pattern)
                )
            )

        query = query.order_by(cls.created_at.desc())

        if limit:
            query = query.limit(limit)

        return query.all()

    @classmethod
    def get_featured_notes(cls, limit=10):
        """Get featured notes"""
        return cls.query.filter_by(is_public=True, is_featured=True)\
                       .order_by(cls.views_count.desc())\
                       .limit(limit).all()

    @classmethod
    def get_popular_notes(cls, limit=10):
        """Get popular notes by views and likes"""
        return cls.query.filter_by(is_public=True)\
                       .order_by((cls.views_count + cls.likes_count).desc())\
                       .limit(limit).all()

    @classmethod
    def find_by_id(cls, note_id):
        """Find note by ID"""
        return cls.query.get(note_id)

    @classmethod
    def get_subjects(cls):
        """Get all unique subjects"""
        subjects = db.session.query(cls.subject.distinct()).all()
        return [subject[0] for subject in subjects]

    def save(self):
        """Save note to database"""
        db.session.add(self)
        db.session.commit()
        return self

    def delete(self):
        """Delete note from database"""
        db.session.delete(self)
        db.session.commit()

    def update(self, **kwargs):
        """Update note attributes"""
        for key, value in kwargs.items():
            if hasattr(self, key):
                setattr(self, key, value)

        self.updated_at = datetime.utcnow()
        db.session.commit()
        return self

File: backend/models/Note.php
<?php
class Note {
    private $conn;
    private $table = "notes";

    public $id;
    public $title;
    public $content;
    public $category_id;
    public $user_id;
    public $is_private;
    public $created_at;
    public $updated_at;

    public function __construct($db) {
        $this->conn = $db;
    }

    public function create() {
        $query = "INSERT INTO " . $this->table . " 
                  SET title=:title, content=:content, category_id=:category_id, 
                      user_id=:user_id, is_private=:is_private";
        
        $stmt = $this->conn->prepare($query);
        
        $this->title = htmlspecialchars(strip_tags($this->title));
        $this->content = htmlspecialchars(strip_tags($this->content));
        $this->category_id = htmlspecialchars(strip_tags($this->category_id));
        $this->user_id = htmlspecialchars(strip_tags($this->user_id));
        $this->is_private = htmlspecialchars(strip_tags($this->is_private));
        
        $stmt->bindParam(":title", $this->title);
        $stmt->bindParam(":content", $this->content);
        $stmt->bindParam(":category_id", $this->category_id);
        $stmt->bindParam(":user_id", $this->user_id);
        $stmt->bindParam(":is_private", $this->is_private);
        
        if ($stmt->execute()) {
            return true;
        }
        
        return false;
    }

    public function read($user_id) {
        $query = "SELECT n.*, c.name as category_name 
                  FROM " . $this->table . " n 
                  LEFT JOIN categories c ON n.category_id = c.id 
                  WHERE n.user_id = :user_id 
                  ORDER BY n.updated_at DESC";
        
        $stmt = $this->conn->prepare($query);
        $stmt->bindParam(":user_id", $user_id);
        $stmt->execute();
        
        return $stmt;
    }

    public function readOne() {
        $query = "SELECT n.*, c.name as category_name 
                  FROM " . $this->table . " n 
                  LEFT JOIN categories c ON n.category_id = c.id 
                  WHERE n.id = :id AND n.user_id = :user_id 
                  LIMIT 0,1";
        
        $stmt = $this->conn->prepare($query);
        $stmt->bindParam(":id", $this->id);
        $stmt->bindParam(":user_id", $this->user_id);
        $stmt->execute();
        
        $row = $stmt->fetch(PDO::FETCH_ASSOC);
        
        if ($row) {
            $this->title = $row['title'];
            $this->content = $row['content'];
            $this->category_id = $row['category_id'];
            $this->is_private = $row['is_private'];
            $this->created_at = $row['created_at'];
            $this->updated_at = $row['updated_at'];
            return true;
        }
        
        return false;
    }

    public function update() {
        $query = "UPDATE " . $this->table . " 
                  SET title=:title, content=:content, category_id=:category_id, 
                      is_private=:is_private 
                  WHERE id=:id AND user_id=:user_id";
        
        $stmt = $this->conn->prepare($query);
        
        $this->title = htmlspecialchars(strip_tags($this->title));
        $this->content = htmlspecialchars(strip_tags($this->content));
        $this->category_id = htmlspecialchars(strip_tags($this->category_id));
        $this->is_private = htmlspecialchars(strip_tags($this->is_private));
        $this->id = htmlspecialchars(strip_tags($this->id));
        $this->user_id = htmlspecialchars(strip_tags($this->user_id));
        
        $stmt->bindParam(":title", $this->title);
        $stmt->bindParam(":content", $this->content);
        $stmt->bindParam(":category_id", $this->category_id);
        $stmt->bindParam(":is_private", $this->is_private);
        $stmt->bindParam(":id", $this->id);
        $stmt->bindParam(":user_id", $this->user_id);
        
        if ($stmt->execute()) {
            return true;
        }
        
        return false;
    }

    public function delete() {
        $query = "DELETE FROM " . $this->table . " 
                  WHERE id = :id AND user_id = :user_id";
        
        $stmt = $this->conn->prepare($query);
        
        $this->id = htmlspecialchars(strip_tags($this->id));
        $this->user_id = htmlspecialchars(strip_tags($this->user_id));
        
        $stmt->bindParam(":id", $this->id);
        $stmt->bindParam(":user_id", $this->user_id);
        
        if ($stmt->execute()) {
            return true;
        }
        
        return false;
    }
}
?>